File: app/src/main/java/com/trevorwiebe/trackacow/data/local/dao/DrugDao.java
package com.trevorwiebe.trackacow.data.local.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.trevorwiebe.trackacow.data.entities.DrugEntity;

import java.util.List;

@Dao
public interface DrugDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertDrug(DrugEntity drugEntity);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertListDrug(List<DrugEntity> drugEntities);

    @Query("SELECT * FROM Drug WHERE drugId = :id")
    DrugEntity getDrugById(String id);

    @Query("SELECT * FROM Drug")
    List<DrugEntity> getDrugList();

    @Query("UPDATE Drug SET defaultAmount = :defaultAmount, drugName = :drugName WHERE drugId = :drugId")
    void updateDrugById(int defaultAmount, String drugName, String drugId);

    @Query("DELETE FROM Drug")
    void deleteDrugTable();

    @Query("DELETE FROM Drug WHERE drugId = :drugId")
    void deleteDrugById(String drugId);

    @Update
    void updateDrug(DrugEntity drugEntity);

    @Delete
    void deleteDrug(DrugEntity drugEntity);

}

File: app/src/main/java/com/trevorwiebe/trackacow/data/local/dao/DrugDao.kt
package com.trevorwiebe.trackacow.data.local.dao

import androidx.room.*
import com.trevorwiebe.trackacow.data.entities.DrugEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface DrugDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertDrug(drugEntity: DrugEntity): Long

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertDrugList(drugList: List<DrugEntity>): List<Long>

    @Query("SELECT * FROM Drug")
    fun getDrugList(): Flow<List<DrugEntity>>

    @Query("UPDATE Drug SET drugName = :drugName, defaultAmount = :defaultAmount WHERE drugPrimaryKey = :id")
    suspend fun updateDrug(drugName: String, defaultAmount: Int, id: Int)

    @Update
    suspend fun updateDrugList(drugList: List<DrugEntity>)

    @Query("DELETE FROM Drug")
    suspend fun deleteAllDrugs()

    @Query("DELETE FROM Drug WHERE drugCloudDatabaseId = :drugId")
    fun deleteDrugById(drugId: String)

    @Delete
    suspend fun deleteDrug(drugEntity: DrugEntity)

    @Transaction
    suspend fun syncCloudDrugListToDatabase(drugList: List<DrugEntity>) {
        deleteAllDrugs()
        insertDrugList(drugList)
    }

}